### DISCUSSION TITLE
Can we skip parent layout in a nested page in app dir?

### DISCUSSION THREAD

**Discussion by surjithctly** [vercel/next.js#47686] [Help] [answered]
### Summary

Imagine I have the following structure.

```
/app
  layout,js (root layout)
  (backend)
  (website)
     page.js
     layout.js (layout for website)
     /about
         /company
            page.js (skip parent for this page)
```

I do not want to use the parent layout in `/about/company/page.js` (Root layout is fine). I want to use a different layout. Also imagine that the website layout is used by many other pages. 

In other words, is there any possibility to send a Props to the Parent Layout? That's fine as well since I can then modify based on it.

### Additional information

```js
Next 13 App Dir
```

**Comment by icyJoseph** [comment, marked as answer]
I think you need to do this: https://beta.nextjs.org/docs/routing/defining-routes#example-opting-specific-segments-into-a-layout

In this case, `/about/company` is placed so that it doesn't get the `(website)` layout, and the `(website)` route group children, are opted into the specific layout.

**Reply by surjithctly** [reply]
So, do I need to have like this? Hmm. But About is part of the Website. It doesn't make any sense to put that outside.

```
(backend)
(website)
(about)
```

I wish there is a better way.

**Reply by ace4port** [reply]
```
(backend)
(website)
  (withLayout)
       page.js
       layout.js (layout for website)
  (about)
```

How about structuring like this?

**Reply by surjithctly** [reply]
Cool, @ace4port this looks like a possible solution. 🙏

**Reply by surjithctly** [reply]
Also, another possible solution I found is using the `usePathname`

```js
"use client";

import { usePathname } from "next/navigation";

export default function GetCondition(props) {
  const pathname = usePathname();
  const condition = pathname === "/about/company";
  return (
    <>{condition ? <RenderAlternate /> : <RenderDefault />}</>
  );
}
```

So I can do the condition based on the pathname, but it only works on client side.

**Reply by gyopiazza** [reply]
Just be aware that with `use client` you're opting-out of RSC for the whole tree.

**Reply by kavinda1995** [reply]
@gyopiazza is it actually? I thought only components that import CC (client components) are become CCs. children of CC don't have to be CC always

**Reply by gyopiazza** [reply]
@kavinda1995 yep, you should also get an error message if you try to import a `'use server'` component from within a `'use client'` tree and it makes sense because server components cannot run in the client.

**Comment by danass** [comment]
> ```
> (backend)
> (website)
>   (withLayout)
>        page.js
>        layout.js (layout for website)
>   (about)
> ```
> 
> How about structuring like this?

isnt that going to create a route hostname/withLayout/(page.js) instead of just root site hostname/?

**Reply by ace4port** [reply]
Since withLayout is for group only, indicated by `( )`, it will not appear in url.

**Reply by giorgi-gvimradze** [reply]
what if we want to skip the `[slug]/layout.tsx` inside `[slug]/[...sub]/page.tsx`

**Reply by julian-castorena-eb** [reply]
Having the same issue where my `blog/layout.tsx` is applied to the `blog/posts/[slug]/ `pages instead of using the layout inside  `blog/posts/[slug]/ `. Nested layouts here is causing trouble, I'll go for the pathname alternative solution in the meantime.

**Reply by growindiedev** [reply]
found any possible solution?

**Reply by julian-castorena-eb** [reply]
I change to group routing, use multiple root layouts where a `page.tsx` is encountered. Work for my use case, hope it helps.

Something like:
```
app/(home)/page.tsx
app/(home)/layout.tsx

app/(blog)/blog/(home)/page.tsx
app/(blog)/blog/(home)/layout.tsx

app/(blog)/blog/posts/[slug]/ page.tsx
app/(blog)/blog/posts/[slug]/ layout.tsx
```

This way only what's inside `(blog)/ ` will share the layout, same for the (home) dir.

**Reply by leadryx97** [reply]
That worked for me as well. My structure looks like this now:

```
app/(posts)/posts/(posts)/layout.js (layout for posts overview page)
-> http://localhost:3000/posts/

app/(posts)/posts/(singlePosts)/[slug]/layout.js (layout for single posts with dynamic routes)
-> http://localhost:3000/posts/post-1
```

**Comment by subhranshudas** [comment]
I have a similar situation -

need common layout for 
```
"/blog"
"/blog/[categorySlug]"
```

but not for 
`"/blog/[categorySlug]/[blogSlug]"`

![Screenshot 2024-04-09 at 2 02 31 PM](https://github.com/vercel/next.js/assets/6158624/65369ecd-a393-40da-a50c-1f00f745557a)


The most nested child page should not have the layout?

**Reply by ace4port** [reply]
probable folder structure

blog
- (withLayout)
  - page.tsx
  - categorySlug
    - page
- [calegorySlug]
  - [blogSlug]
    - page

**Reply by subhranshudas** [reply]
do you mean like this?

```
blog
     (withLayout)
                page.tsx
                [categorySlug]
                         page.tsx
      [calegorySlug]
                [blogSlug]
                         page.tsx
```

when i hit "/blog/cat1" how does Next JS knows which route group to pick from?

**Reply by subhranshudas** [reply]
oh i see, the 2nd one does not have a `page.tsx` at the `[categorySlug]` level?

**Reply by leadryx97** [reply]
It is not the exact same case but I made two different layouts work with this structure:

```
app/(posts)/posts/(posts)/layout.js (layout for posts overview page)
-> http://localhost:3000/posts/

app/(posts)/posts/(singlePosts)/[slug]/layout.js (layout for single posts with dynamic routes)
-> http://localhost:3000/posts/post-1
```

Maybe it helps.

**Comment by adnan-smlatic** [comment]
Any updates on this in 2024?

I am trying to have a completely separate layout for my (checkout) cart route, but it keeps adding the navbar and footer from the root layout. None of the above solutions have worked for me.

```
app
├── [locale]
│   ├── (checkout)
│   │   ├── cart
│   │   │   └── page.tsx
│   │   └── layout.tsx
│   ├── [category]
│   │   ├── [product]
│   │   │   └── page.tsx
│   │   └── page.tsx
│   ├── blog
│   │   └── [slug]
│   ├── layout.tsx
│   ├── not-found.tsx
│   ├── page.tsx
│   ├── providers.tsx
```

**Reply by anurag629** [reply]
Same problem I'm also facing. Did you find any solution?

**Reply by adnan-smlatic** [reply]
I haven't tried it, but it seems like the only way is:

- [locale]
- - (layout1)
- - - layout.tsx
- - (layout2)
- - - layout2.tsx


Although I don't understand how the root route works in this case. 

Either way this entire thing doesn't make any sense. Nested layouts should overwrite the parent and optionally Extend it.

**Comment by danass** [comment]
I guess you should create two folders, one (withlayout) and one (nolayout)
at root level

Le mar. 6 août 2024 à 16:16, Adnan Smlatić ***@***.***> a
écrit :

**Comment by KrownWealth** [comment]
Route groups work well for me in this scenario I wanted to have a different root layout for my dashboard so I group like this below 

![group-route](https://github.com/user-attachments/assets/f8b71058-b38e-4004-b3bb-d8ffd0748bad)

**Comment by thefomin** [comment]
You may find this solution useful

Even though routes inside (filter) and (page) share the same URL hierarchy:
/filter
/about
/faq
...
you can create a different layout for each group by adding a layout.js file inside their folders.

<img width="447" alt="image" src="https://github.com/user-attachments/assets/9867b760-9df8-453e-84af-4b6fc2e1bf47">



docs:
https://nextjs.org/docs/app/building-your-application/routing/route-groups#convention

**Comment by EfstathiadisD** [comment]
I have a question re. this so Inam gonna post it here:

I have the following need:
- organizations route (LAYOUT 1)
- - [oid] (More pages under here with just LAYOUT 1)
- - new (NO LAYOUT, can be intercepted in multiple levels)
- - - applications route (LAYOUT 2)
- - - - new (NO LAYOUT, can be intercepted in multiple levels)
- - - - [aid] (More pages under here with just LAYOUT 2)
- - - stores route (LAYOUT 3)
- - - - new (NO LAYOUT, can be intercepted in multiple levels)
- - - - [sid] (More pages under here with just LAYOUT 3)

Hope the intention is clear. I can use pathname but it will be messy as hell and it feels like a hack. How would you structure it? I need some help.

**Reply by EfstathiadisD** [reply]
I guess I can have a route groups (organizations) which will have just the organizations route and its direct sub routes.

Then have an (applications) route groups which will have organizations/[oid] and the direct routes of it.

And so on for the stores. Will try that and again feels like a hack. Or not? Don't know. File based routing is freakish for such scenarios.

**Reply by EfstathiadisD** [reply]
Here is what worked.. Not sure if there is a better way but I think it's a good one.. It would be nice if @leerob or smn from the NextJS Team could share some inshights. For now it works and I kind of like it..
```
apps/merchant-ui/src/app/
├── [locale]
│   ├── (auth)
│   │   └── login
│   │       ├── components
│   │       │   ├── Advertorial
│   │       │   │   └── Advertorial.tsx
│   │       │   ├── Authenticate
│   │       │   │   └── Authenticate.tsx
│   │       │   └── Form
│   │       │       └── Form.tsx
│   │       ├── constants
│   │       │   ├── schema.ts
│   │       │   └── urls.ts
│   │       ├── graphql
│   │       │   └── queries.graphql
│   │       ├── locales
│   │       │   ├── el-GR.json
│   │       │   ├── en-GB.json
│   │       │   └── en-US.json
│   │       └── page.tsx
│   ├── future
│   │   ├── (account)
│   │   │   └── [aid]
│   │   │       └── settings
│   │   │           ├── authentication
│   │   │           │   └── page.tsx
│   │   │           └── contact
│   │   │               └── page.tsx
│   │   ├── (applications)
│   │   │   └── organizations
│   │   │       └── [oid]
│   │   │           └── applications
│   │   │               ├── [aid]
│   │   │               │   ├── banking
│   │   │               │   │   └── page.tsx
│   │   │               │   ├── business
│   │   │               │   │   └── page.tsx
│   │   │               │   ├── contact
│   │   │               │   │   └── page.tsx
│   │   │               │   └── review
│   │   │               │       └── page.tsx
│   │   │               ├── layout.tsx
│   │   │               └── page.tsx
│   │   ├── (organizations)
│   │   │   └── organizations
│   │   │       ├── [oid]
│   │   │       │   ├── members
│   │   │       │   │   └── page.tsx
│   │   │       │   └── settings
│   │   │       │       └── page.tsx
│   │   │       ├── layout.tsx
│   │   │       └── page.tsx
│   │   └── (stores)
│   │       └── organizations
│   │           └── [oid]
│   │               └── stores
│   │                   ├── [sid]
│   │                   │   └── home
│   │                   │       └── page.tsx
│   │                   ├── layout.tsx
│   │                   └── page.tsx
│   └── layout.tsx
├── global-error.tsx
├── globals.css
└── robots.ts

37 directories, 29 files
```

What we get:
- Organizations, Applications and Stores get their own unique Layouts. (Some speed issues but OK)
- Top Level Layout is kept and all the Providers are inherited. (I was scared about that)
- No confusion of the routes as long as we don't have a `page.tsx`.


The more I think about it the more I find it tastefull. Not sure. It would be way easier with code but it works..
